### PR TITLE
Allow creating ClusterRoleBinding with parameters, and adding rules to ClusterRoles

### PR DESCRIPTION
**Pull request by AdiZav**

### CODE REVIEW

**Comment by AdiZav** [comment]
!verified

**Comment by AdiZav** [comment]
!verified

File: ocp_resources/cluster_role_binding.py [hunk 1]
```diff
@@ -9,3 +9,27 @@ class ClusterRoleBinding(Resource):
...
+        self.subjects = subjects
+        self.res = None
+
+    def to_dict(self):
+
+        self.res = super().to_dict()
+        self.res.setdefault("roleRef", {})
+        self.res["roleRef"] = {
+            "apiGroup": self.api_group,
+            "kind": "ClusterRole",
```
Comment (myakove) [inline ocp_resources/cluster_role_binding.py:30]:
```suggestion
            "kind": ClusterRole.kind,
```

**Review by myakove — CHANGES_REQUESTED** [review]

File: ocp_resources/cluster_role_binding.py (hunk 1 above)
Comment (AdiZav) [reply ocp_resources/cluster_role_binding.py:30]: ACK

File: ocp_resources/cluster_role_binding.py [hunk 2]
```diff
@@ -9,3 +9,27 @@ class ClusterRoleBinding(Resource):
     """
 
     api_group = Resource.ApiGroup.RBAC_AUTHORIZATION_K8S_IO
+
+    def __init__(
+        self,
+        name=None,
+        cluster_role=None,
+        subjects=[],
```
[the remark this answers is not in the document]
Comment (AdiZav) [reply ocp_resources/cluster_role_binding.py:17]: ACK

**Comment by AdiZav** [comment]
!verified

File: ocp_resources/cluster_role_binding.py [hunk 3]
```diff
@@ -9,3 +10,27 @@ class ClusterRoleBinding(Resource):
...
+        super().__init__(name=name)
+        self.cluster_role = cluster_role
+        self.subjects = subjects
+        self.res = None
+
+    def to_dict(self):
+
+        self.res = super().to_dict()
+        self.res.setdefault("roleRef", {})
+        self.res["roleRef"] = {
```
Comment (myakove) [inline ocp_resources/cluster_role_binding.py:29]: cluster_role and subjects can be None, don't you need to check before set them?
Is this valid role without them?

File: ocp_resources/cluster_role_binding.py [hunk 4]
```diff
@@ -9,3 +10,27 @@ class ClusterRoleBinding(Resource):
...
+    def __init__(
+        self,
+        name=None,
+        cluster_role=None,
+        subjects=None,
+    ):
+        super().__init__(name=name)
+        self.cluster_role = cluster_role
+        self.subjects = subjects
+        self.res = None
```
Comment (myakove) [inline ocp_resources/cluster_role_binding.py:23]:
```
        self.res = None
```
Why this is needed?

File: ocp_resources/cluster_role.py [hunk 5]
```diff
@@ -33,19 +33,53 @@ def __init__(
...
+        if rule:
+            self.set_rule(rule=rule)
+
+        return self.res
+
+    def set_rule(self, rule):
+        if not self.res:
+            self.res = super().to_dict()
+        # Drop the rule if it's already in the list
+        rules = [
```
Comment (myakove) [inline ocp_resources/cluster_role.py:78]: multiple rules for the same resource are not valid?

File: ocp_resources/cluster_role.py [hunk 6]
```diff
@@ -33,19 +33,53 @@ def __init__(
...
+    def add_rule(
+        self,
+        api_groups=None,
+        permissions_to_resources=None,
+        verbs=None,
+    ):
+        if not self.res:
+            self.res = self.to_dict()
+
+        rule = {}
```
Comment (myakove) [inline ocp_resources/cluster_role.py:62]: role

File: ocp_resources/cluster_role.py [hunk 7]
```diff
@@ -33,19 +33,53 @@ def __init__(
...
-        if self.api_groups:
-            rules["apiGroups"] = self.api_groups
         if self.permissions_to_resources:
-            rules["resources"] = self.permissions_to_resources
-        if self.verbs:
-            rules["verbs"] = self.verbs
-        if rules:
-            res["rules"] = [rules]
-        return res
+            self.res = self.add_rule(
```
Comment (myakove) [inline ocp_resources/cluster_role.py:45]: you override self.res here.

**Review by myakove — CHANGES_REQUESTED** [review]

File: ocp_resources/cluster_role_binding.py (hunk 4 above)
Comment (AdiZav) [reply ocp_resources/cluster_role_binding.py:23]: To allow adding rules after initiating object

File: ocp_resources/cluster_role_binding.py (hunk 3 above)
Comment (AdiZav) [reply ocp_resources/cluster_role_binding.py:29]: cluster_role required, setting to mandatory attribute. 
Subjects aren't required, will add check

File: ocp_resources/cluster_role.py [hunk 8]
```diff
@@ -33,19 +33,46 @@ def __init__(
...
+        if permissions_to_resources:
+            rule["resources"] = permissions_to_resources
+        if verbs:
+            rule["verbs"] = verbs
+        if rule:
+            self.set_rule(rule=rule)
+
+        return self.res
+
+    def set_rule(self, rule):
```
Comment (myakove) [inline ocp_resources/cluster_role.py:73]: Is there any point to keep this public? User can call set_rule()? 
I think it should be private, and then you can remove 
```
if not self.res:
            self.res = super().to_dict()
```

**Comment by myakove** [comment]
@AdiZav please rebase.

File: ocp_resources/cluster_role.py [hunk 9]
```diff
@@ -33,19 +33,46 @@ def __init__(
         self.api_groups = api_groups
         self.permissions_to_resources = permissions_to_resources
         self.verbs = verbs
+        self.desired_state = {"rules": []}
 
     def to_dict(self):
-        res = super().to_dict()
+        self.res = super().to_dict()
```
Comment (rnetser) [inline ocp_resources/cluster_role.py:39]: if you set a value to setf.res here, you need to define it in __init__, do not hide it here

File: ocp_resources/cluster_role.py (hunk 8 above)
Comment (rnetser) [reply ocp_resources/cluster_role.py:73]: +1

File: ocp_resources/cluster_role_binding.py [hunk 10]
```diff
@@ -9,3 +10,33 @@ class ClusterRoleBinding(Resource):
...
+        self.cluster_role = cluster_role
+        self.subjects = subjects
+
+    def to_dict(self):
+
+        res = super().to_dict()
+        res.setdefault("roleRef", {})
+        if self.cluster_role:
+            """
+            It's the responsibility of the caller to verify the desired configuration they send.
```
Comment (rnetser) [inline ocp_resources/cluster_role_binding.py:30]: This comment is not needed; this is applicable to almost anything passed by callers to the api

**Review by rnetser — CHANGES_REQUESTED** [review]

File: ocp_resources/cluster_role_binding.py (hunk 10 above)
Comment (AdiZav) [reply ocp_resources/cluster_role_binding.py:30]: ACK

File: ocp_resources/cluster_role.py (hunk 9 above)
Comment (AdiZav) [reply ocp_resources/cluster_role.py:39]: Defined in Resource.init()

File: ocp_resources/cluster_role_binding.py [hunk 11]
```diff
@@ -9,3 +10,28 @@ class ClusterRoleBinding(Resource):
...
+        cluster_role=None,
+        subjects=None,
+    ):
+        super().__init__(name=name)
+        self.cluster_role = cluster_role
+        self.subjects = subjects
+
+    def to_dict(self):
+
+        res = super().to_dict()
```
Comment (myakove) [inline ocp_resources/cluster_role_binding.py:26]: let's use here self.res as well
We need to keep it consistent and we will need to adjust all other classes.

**Review by myakove — CHANGES_REQUESTED** [review]

File: ocp_resources/cluster_role_binding.py (hunk 11 above)
Comment (AdiZav) [reply ocp_resources/cluster_role_binding.py:26]: Ok

File: ocp_resources/cluster_role.py (hunk 9 above)
Comment (rnetser) [reply ocp_resources/cluster_role.py:39]: ack

**Comment by AdiZav** [comment]
!verified

File: ocp_resources/cluster_role_binding.py [hunk 12]
```diff
@@ -9,3 +10,28 @@ class ClusterRoleBinding(Resource):
...
+        subjects=None,
+    ):
+        super().__init__(name=name)
+        self.cluster_role = cluster_role
+        self.subjects = subjects
+
+    def to_dict(self):
+
+        self.res = super().to_dict()
+        self.res.setdefault("roleRef", {})
```
Comment (myakove) [inline ocp_resources/cluster_role_binding.py:27]: `self.res.setdefault("roleRef", {})`
Is this always needed? if not move it under the if.

**Review by myakove — CHANGES_REQUESTED** [review]

File: ocp_resources/cluster_role.py [hunk 13]
```diff
@@ -33,19 +33,45 @@ def __init__(
...
-        if self.api_groups:
-            rules["apiGroups"] = self.api_groups
-        if self.permissions_to_resources:
-            rules["resources"] = self.permissions_to_resources
-        if self.verbs:
-            rules["verbs"] = self.verbs
-        if rules:
-            res["rules"] = [rules]
-        return res
+        if not self.res:
```
[the remark this answers is not in the document]
Comment (AdiZav) [reply ocp_resources/cluster_role.py]: ack

**Comment by AdiZav** [comment]
!verified

**Comment by AdiZav** [comment]
!verified

**Review by myakove — APPROVED** [review]

**Review by rnetser — APPROVED** [review]

**Pull request merged**